misk/src/main/kotlin/misk/flags/DynamicFlagModule.kt

```diff
@@ -0,0 +1,163 @@
+            prefix: String = "",
+            qualifier: Annotation? = null
+    ) {
+        val actualQualifier = qualifier ?: if (!prefix.isEmpty()) Names.named(prefix) else null
+        val constructor = kclass.constructors.firstOrNull {
+            it.parameters.size == 1 && it.parameters[0].type.classifier == Flags.Context::class
+        } ?: throw IllegalArgumentException(
+                "$kclass has no single argument constructor taking a Flags.Context")
+
+
+        if (actualQualifier == null) {
+            bind(kclass.java)
+                    .toProvider(FlagsProvider(constructor, prefix))
+                    .asEagerSingleton()
+        } else {
+            bind(kclass.java)
+                    .annotatedWith(actualQualifier)
+                    .toProvider(FlagsProvider(constructor, prefix))
+                    .asEagerSingleton()
+        }
+    }
+
+    fun bindBooleanFlag(name: String, desc: String, qualifier: Annotation = Names.named(name)) {
+        bind<BooleanFlag>()
+                .annotatedWith(qualifier)
+                .toProvider(BooleanFlagProvider(name, desc))
+                .asEagerSingleton()
+    }
+
+    fun bindStringFlag(name: String, desc: String, qualifier: Annotation = Names.named(name)) {
+        bind<StringFlag>()
+                .annotatedWith(qualifier)
+                .toProvider(StringFlagProvider(name, desc))
+                .asEagerSingleton()
+    }
+
+    fun bindIntFlag(name: String, desc: String, qualifier: Annotation = Names.named(name)) {
+        bind<IntFlag>()
+                .annotatedWith(qualifier)
+                .toProvider(IntFlagProvider(name, desc))
+                .asEagerSingleton()
+    }
+
+    fun bindDoubleFlag(name: String, desc: String, qualifier: Annotation = Names.named(name)) {
+        bind<DoubleFlag>()
+                .annotatedWith(qualifier)
+                .toProvider(DoubleFlagProvider(name, desc))
+                .asEagerSingleton()
+    }
+
+    inline fun <reified T : Any> bindJsonFlag(
+            name: String,
+            desc: String,
+            qualifier: Annotation = Names.named(name)
+    ) = bindJsonFlag(T::class, name, desc, qualifier)
+
+    fun <T : Any> bindJsonFlag(
+            kclass: KClass<T>,
+            name: String,
+            desc: String,
+            a: Annotation = Names.named(name)
+    ) {
+        val flagType = parameterizedType<JsonFlag<T>>(kclass.java)
+
+        @Suppress("UNCHECKED_CAST")
+        val flagTypeLiteral = TypeLiteral.get(flagType) as TypeLiteral<JsonFlag<T>>
+        bind(flagTypeLiteral)
+                .annotatedWith(a)
+                .toProvider(JsonFlagProvider(kclass, name, desc))
+                .asEagerSingleton()
+    }
+
+    private class BooleanFlagProvider(val name: String, val desc: String) : Provider<BooleanFlag> {
+        @Inject
+        lateinit var flagStore: DynamicFlagStore
+
+        override fun get(): BooleanFlag = flagStore.registerBooleanFlag(name, desc)
```

Comment:
`register` is an interesting word because that’s lazy, right?

Should we publish the description before the flag’s value is requested?

Reply:
Yes, as documented on the flag store, register should be an asynchronous, and only block until the value is retrieved. We can make description a required blocking operation as well, although I'm generally fine serving description out of local cache prior to whatever backend server needs to be reached has become available.